napi-inl.h

```diff
@@ -322,6 +322,33 @@ inline Value Env::RunScript(String script) {
   return Value(_env, result);
 }
 
+#ifdef NAPI_EXPERIMENTAL
+template <typename T>
+inline T* Env::AddonData() {
+  void* data = nullptr;
+  napi_status status = napi_get_instance_data(_env, &data);
+  NAPI_THROW_IF_FAILED(_env, status, nullptr);
+
+  if (data == nullptr) {
+    T* new_data = new T();
```

Comment:
Do we maybe want to store `type_index(typeid(T))` on the `Env` class and verify that the value remains identical in subsequent calls to this method, for proper type safety?

Reply:
Now that I know you can do that, definitely 😀

Reply:
@gabrielschulhof I think `typeid` is only going to work in the exception-enabled variant, though, because otherwise RTTI is not enabled … you may have to play around with that a bit

Reply:
@addaleax we can document that the type safety of the instance data is tied to the use of exceptions. It's not 100% safe either, because one can always use `napi_set_instance_data()` to completely pull the rug out from under the whole node-addon-api addon.

I guess there's an assumption here that the codebase for the addon is monolithic in terms of who controls it.

Reply:
Yeah, and to be clear, my goal would be to prevent accidental mistakes, not being completely fool-proof :)

Reply:
@addaleax both exception-using and non-exception-using versions are currently built using `-fno-rtti`. Thus, I'm not sure that we can add this, lest we require yet another thing for people to add to their `binding.gyp`s.

Reply:
Was that always the case? I kind of assumed that `-frtti` was required to properly deal with subclassed exceptions (e.g. catching an instance of `class A : public B {}` with `catch (B& exception)`)?

Reply:
AFAICT it was always the case.